### DISCUSSION TITLE
FadeIn opacity

### DISCUSSION THREAD

**Discussion by leeenglestone** [StereoKit/StereoKit#387] [Ideas]
If there isn't a way already, maybe consider a way of fading in opacity from 0 to 1?

**Comment by maluoi** [comment]
I might like to know a bit more about the context of this, what category of thing in particular you're trying to fade in and out would be quite interesting to me!

The general requirements for fading something in and out are a `Material` with `material.Transparency = Transparency.Blend`, and an alpha value in either `material[MatParamName.ColorTint]` or (generally better) the color value you pass into `Draw`.

In general, since StereoKit does try not to store much state, there's not a ton of opportunities for SK to do the fading for you! Maybe some with respect to UI? I have considered providing some sort of tweening library to make that category of activity easier, but would love to know your thoughts here :)

**Reply by leeenglestone** [reply]
I'm led to believe that there are a number of things in UI and interfaces that are jarring to the brain because the are not found in nature.
Things like
- The pure colour black (artificial vantablack aside)
- Things that just appear/disappear in the blink of an eye
- Thanks that go from stationary to final speed without acceleration (and vice versa)

I personally feel allowing something fade into or out of existence over a second or half a second is more natural than a sudden absolute appear/disappear. But I know not of the underlying reasons why this is difficult due to state so I believe you when you say it is difficult/not been accommodated for. I wouldn't be surprised if this was increasingly requested however.

**Reply by maluoi** [reply]
Okay! I'm glad this is about UI, because this is one of the few places where I _do_ have some opportunity for this type of thing :) I've also got some UI refactors in-progress right now, and this is very much the sort of feature I'm hoping to account for!

I did recently find a way to add animation to the immediate mode paradigm, and this is visible to some degree on buttons right now! But not really everywhere, it's still a bit tricky. I've since learned a bunch of other new tricks though, and should be able to do a lot more with the next iteration of the UI's architecture. Definitely looking forward to shipping some of that work!

My current UI goal is to recreate the look of MRTK3's MRDL visual style. I work with a lot of those people, and it's an excellent target! Being able to implement that should round out StereoKit's UI look and feel capabilities quite nicely :)

And yeah, black is also not great for additive displays like HoloLens :) There shouldn't be any black UI in SK!
